standard/ranges.md

```diff
@@ -0,0 +1,291 @@
+    }
+```
+
+`Index` values are constructed from an `int`, specifying the positive offset, and a `bool`, indicating whether the offset is from the end (`true`) or start (`false`). If the specified offset is negative an `ArgumentOutOfRangeException` is thrown.
+
+> *Example*
+>
+> ```CSharp
+> Index first = new Index(0, false); // first element index
+> var last = new Index(1, true);     // last element index
+> var past = new Index(0, true);     // past-end index
+>
+> Index invalid = new Index(-1);     // throws ArgumentOutOfRangeException
+> ```
+>
+> *end example*
+
+There is an implicit conversion from `int` to `Index` which produces from-start indicies, and a language-defined unary operator `^` (§hat-operator) from `int` to `Index` which produces from-end indicies.
+
+> *Example*
+>
+> Using implicit conversions and the unary `^` operator the above examples may be written:
+>
+> ```CSharp
+> Index first = 0; // first element index
+> var last = ^1;   // last element index
+> var past = ^0;   // past-end index
+> ```
+>
+> *end example*
+
+The method `GetOffset` converts from an abstract `Index` value to a concrete `int` index value for a sequence of the specified `length`.
+
+If the `Index` value, `I`, is from-end this method returns the same value as `length - I.Value`, otherwise it returns the same value as `I.Value`.
+
+This method does **not** check that the return value is in the valid range of `0` through `length-1` inclusive.
+
+> *Note:* No checking is specified as the expected use of the result is to index into a sequence with `length` elements, and that indexing operation is expected to perform the appropriate checks. *end note*
+
+`Index` implements `IEquatable<Index>` and values may be compared for equality. However `Index` values are not ordered and no other comparison operations are provided.
+
+> *Note:* `Index` values are unordered as they are abstract indicies, it is in general impossible to determine whether a from-end index comes before or after a from start index without reference to a sequence length. Once converted to concrete indicies, e.g. by `GetOffset`, those concrete indicies are comparable. *end note*
+
+`Index` values may be directly used in the *argument_list* of an *element_access* expression (§12.8.12):
+
+- which is statically bound (§12.3.2) and:
+  - it is an array access and the target is a single-dimensional array (§12.8.12.2);
+  - it is a string access (§string-access); or
+  - it is an indexer access and the target type conforms to a sequence pattern for which implicit `Index` support is specified (§24.4.2).
+- which is statically or dynamically bound (§12.3.2) and:
+  - it is an indexer access and the target type has an indexer with parameters of `Index` type (§12.8.12.3);
+
+## 24.3 The `Range` type
+
+The `System.Range` type represents the abstract range of `Index`es from a `Start` index up to, but not including, an `End` index.
+
+```CSharp
+    public readonly struct Range : IEquatable<Index>
```

Comment:
This type declaration duplicates what is proposed (and needed) in C.3. Let's replace this one with a link to C.3.

Reply:
See comment on `Index`